### ISSUE TITLE
Explicit node selection for a job

### ISSUE THREAD

**Issue by zhiltsov-max** [microsoft/pai#2878]
The issue is for a proposal and discussion.

**What would you like to be added**:
Support for explicit node selection for a job and, probably, logical cluster segmentation or reservations.

**Why is this needed**:
Imagine there is a single cluster and a few teams using it. The cluster has several logical parts: a part for each team, a common part and, may be, a testing/dedicated segment. Mainly, the teams are happy with their segments and the common one, but sometimes really BIG tasks should be executed, consuming all the resources. The cluster is not very big, so teams tend to do manual resource management on their partitions by agreement. 
So, the cluster has logical parts, but should not be split completely. A node can be attached to multiple parts. It is of average size and has high utilization. 

**Without this feature, how does the current module work**:
- GPU type can be set to be unique for each node, so that users could select the right node. While people run single-node tasks it is ok, but If a user selects no GPU type, a task can be scheduled to a "reserved" node (including other team's one). With distributed tasks we can ether launch a "single-node distributed task" or we can launch a really distributed task with no control of target nodes, while task can be tailored for, to say, specific GPU properties.
- Virtual clusters sounds like the right thing to do at least for cluster segmentation, but it seems it can only be split by load, without specific nodes assignment. It's unclear how to use this properly for GPUs. They won't allow to launch whole cluster tasks as tasks can only be put to a single Hadoop queue.
- Hadoop node labels are unique and look like no more than host names. Probably, it could help with single-node tasks scheduling, but we still need some changes for distributed tasks.
- Hadoop node filters. I haven't dug this deep, but they're based on labels.

**What is proposed**
- Add (arbitrary) labels for nodes, e.g. on-board GPUs, CPUs, memory, subcluster(-s). They are probably could be set in `layout.yaml`. Some, related to resources, could be populated automatically.
- Change the job description so that it contains a list of acceptable nodes / filters instead of / in addition to virt. cluster & GPU type. Scheduler will use this list for task placement. Provide meta-label "any".
- Provide users a new job submission UI element (an image follows) to help users with node selection by providing filters, based on node labels.
- Add user property "team appearance" to provide default node selection
- Add node access user rights checking in backend on job submission

![sketch1](https://user-images.githubusercontent.com/13832349/58814926-4f16f400-862f-11e9-8432-59ee449d819f.png)
Filters are similar to GitLab's or GitHub's issue labels. Labels are similar to K8s ones.

**Components that may involve changes**:
- `framework launcher`. Mostly - application master's `SelectionManager`.
- `rest-server`, `framework launcher` - job description
- `webportal` - node selection UI element

**Comment by abuccts** [comment]
Hi, thanks for your proposal! This feature is worked in progress #2770.

/cc @mzmssg Could you please post the current design in this thread for discussion?

**Comment by scarlett2018** [comment]
@zhiltsov-max - you are very welcome to join the design and implementation for #2770 .

**Comment by mzmssg** [comment]
@abuccts 
It looks like the extension of GPU type, not the existing issue.

@zhiltsov-max 
Thank you for your suggestion, which could make scheduler flexible. I want to share some thoughts about backend:
1. About yarn node label
Yarn node label is to partition resource. Of course, we could make per node a partition, then assign task to specific nodes. But as you said, it's no more than hostname, the only benefic is YARN will validate the accessment. 
Another option is adding node attribute in node-label, but unlike k8s label, yarn node label is exclusive, every node could only be assgined one label, and submission should exactly match the label. So we can't make yarn aware of a rich constraint expression, such as "CPU=Xeon E5, GPU=*". So it might **not** be a proper solution.
2. About virtual cluster
Actually, combined with node-label, virtual cluster could split resource by nodes and load(and do validation). Defect see 1.
3. About GPU type
In my opinion, GPU type is a workaround for yarn label: yarn can't provide a expected scheduling, then we do it in launcher, or called SelectionManager, yarn can't provide a rich constraint, then we add it outside, or called GPU type. Of cource we could customize more labels and logic in Launcher, it could work, but we might not leaverage yarn's feature to do more.
4. Hadoop node fitler
I'm not clear about it. Do you mean the [placementConstraint](https://hadoop.apache.org/docs/r3.1.1/hadoop-yarn/hadoop-yarn-site/PlacementConstraints.html) in hadoop3?  On yarn, It might be the better solution, the only trouble is the upgrading.

So basically, on current implementation, I agree that adding such item into launcher and filter in SelectionManager are the best solution.

But according to our plan, we prefer to migrate to k8s, make PAI a pure k8s project firstly, then based on k8s to do more feature. After that, we might find more platform support for your proposal.

**Comment by zhiltsov-max** [comment]
1. VC's do not seem as the right solution, unless you're going to create a VC for each task. They can restrict nodes in the case of a few teams, but there is still the problem with "reservations" and manual resource distribution.

2. More thoughts on the topic. SLURM scheduler provides node whitelist, node blacklist, and cluster partitions at the same time. They can be specified at the task submission time. May be PAI should take a look in this direction and just extend task submission options with these lists.

3. Hadoop 2 provides "node label expressions" described here:
    - Section "Specifying node label for application" 
        https://hadoop.apache.org/docs/r2.7.3/hadoop-yarn/hadoop-yarn-site/NodeLabel.html
    - http://hadoop.apache.org/docs/r2.8.2/api/org/apache/hadoop/yarn/api/records/ApplicationSubmissionContext.html#getNodeLabelExpression() and set in ctor.

**Comment by mzmssg** [comment]
@zhiltsov-max 
Let's forget #2770, which is to provide a better inter vc isolation, not for task placement.

> Hadoop node labels are unique and look like no more than host names. Probably, it could help with single-node tasks scheduling, but we still need some changes for distributed tasks.

> Hadoop node filters. I haven't dug this deep, but they're based on labels.

1. Actually I'm confused about the difference of these two solutions. My understanding: 
   1. the first one is  [Node label](https://hadoop.apache.org/docs/r2.9.0/hadoop-yarn/hadoop-yarn-site/NodeLabel.html), 
   2. the second is [PlacementConstraint](https://hadoop.apache.org/docs/r3.1.1/hadoop-yarn/hadoop-yarn-site/PlacementConstraints.html) and  [Node attribute](https://hadoop.apache.org/docs/r3.2.0/hadoop-yarn/hadoop-yarn-site/NodeAttributes.html). 
 Please correct me.
2. NodeLabel is not enough, that's why we need node attribute.

3. Allow manual scheduling might break some scheduler assumption, it's a tradeoff between flexibility and performance. 

+ @scarlett2018 to evaluate this request.

**Comment by zhiltsov-max** [comment]
@mzmssg 
OK, by just "labels" I meant "single label matching" and by "filters" I meant "label expressions". They indeed belong to a single approach, you're right.

Hadoop "Node attributes are look very close to the thing I described by arbitrary node labels in the proposal. The only problem with attribute-based filtration is that after filtration we can end up with a set of nodes, which contain a few undesired nodes. That's why node blacklist and/or whitelist may be required to complete the solution.

**Comment by mzmssg** [comment]
@zhiltsov-max 
Yeah, luckly, YARN have it.
http://hadoop.apache.org/docs/r2.9.0/api/org/apache/hadoop/yarn/api/records/ResourceBlacklistRequest.html

**Comment by zhiltsov-max** [comment]
Something related mentioned in discussion of #445 and in #2062. The UI concept above can be used for this as well.

**Comment by fanyangCS** [comment]
I think we support this in HiveD.
